api/pkg/service/admin/admin.go

```diff
@@ -3,34 +3,172 @@ package admin
 import (
 	"context"
 	"fmt"
+	"strings"
 
+	"github.com/dgrijalva/jwt-go"
 	"github.com/jinzhu/gorm"
 	"go.uber.org/zap"
-	"goa.design/goa/v3/security"
 
 	"github.com/tektoncd/hub/api/gen/admin"
 	"github.com/tektoncd/hub/api/pkg/app"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"github.com/tektoncd/hub/api/pkg/service/auth"
+	"github.com/tektoncd/hub/api/pkg/token"
 )
 
 type service struct {
+	*auth.Validator
 	logger *zap.SugaredLogger
 	db     *gorm.DB
 	jwtKey string
 }
 
+var (
+	invalidTokenError  = admin.MakeInvalidToken(fmt.Errorf("invalid user token"))
+	invalidScopesError = admin.MakeInvalidScopes(fmt.Errorf("user not authorized"))
+	internalError      = admin.MakeInternalError(fmt.Errorf("failed to create agent"))
+)
+
 // New returns the admin service implementation.
 func New(api app.Config) admin.Service {
-	return &service{api.Logger(), api.DB(), api.JWTSigningKey()}
-}
-
-// JWTAuth implements the authorization logic for service "admin" for the "jwt"
-// security scheme.
-func (s *service) JWTAuth(ctx context.Context, token string, scheme *security.JWTScheme) (context.Context, error) {
-	return ctx, fmt.Errorf("not implemented")
+	return &service{
+		Validator: &auth.Validator{
+			DB:     api.DB(),
+			Logger: api.Logger().With("service", "validator"),
+			JWTKey: api.JWTSigningKey(),
+		},
+		logger: api.Logger(),
+		db:     api.DB(),
+		jwtKey: api.JWTSigningKey(),
+	}
 }
 
 // Create a agent user with required scopes
 func (s *service) CreateAgent(ctx context.Context, p *admin.CreateAgentPayload) (res *admin.CreateAgentResult, err error) {
-	s.logger.Info("admin.CreateAgent")
-	return
+
+	tx := s.db.Begin()
+
+	// Check if a normal user exists with the name in payload
+	user := &model.User{}
+	q := tx.Model(&model.User{}).
+		Where("LOWER(github_name) = ?", strings.ToLower(p.Name))
+
+	if err = q.Find(&user).Error; err == nil {
+		return nil, admin.MakeInvalidPayload(fmt.Errorf("normal user exists with name: %s", p.Name))
```

Comment:
Don't we need to check the scopes also here ??

Reply:
We do check if the scopes exist but in the latter part. here we check if a user exists with the agent's name.